File: frontend/src/services/recommendationService.ts
import { apiClient } from './api';
import { ApiResponse } from '../types/api';
import {
  RecommendationItem,
  UserPreferenceProfile,
  TrackActivityPayload,
  SubmitFeedbackPayload,
  RecommendationFeedbackType,
} from '../types/recommendation';

export const recommendationService = {
  /**
   * Get personalized recommendations for user (mixed flights + hotels + destinations).
   */
  async getRecommendations(
    limit: number = 8,
    context?: string,
    destination?: string
  ): Promise<RecommendationItem[]> {
    const response = await apiClient.get<ApiResponse<RecommendationItem[]>>('/v1/recommendations', {
      params: { limit, context, destination },
    });
    return response.data.data || [];
  },

  /**
   * Get flight recommendations with explainability.
   */
  async getFlightRecommendations(limit: number = 6): Promise<RecommendationItem[]> {
    const response = await apiClient.get<ApiResponse<RecommendationItem[]>>('/v1/recommendations/flights', {
      params: { limit },
    });
    return response.data.data || [];
  },

  /**
   * Get hotel recommendations with explainability.
   */
  async getHotelRecommendations(limit: number = 6): Promise<RecommendationItem[]> {
    const response = await apiClient.get<ApiResponse<RecommendationItem[]>>('/v1/recommendations/hotels', {
      params: { limit },
    });
    return response.data.data || [];
  },

  /**
   * Get personalized and trending destinations.
   */
  async getDestinationRecommendations(limit: number = 6): Promise<RecommendationItem[]> {
    const response = await apiClient.get<ApiResponse<RecommendationItem[]>>('/v1/recommendations/destinations', {
      params: { limit },
    });
    return response.data.data || [];
  },

  /**
   * Get popular destinations (public alias).
   */
  async getPopularDestinations(limit: number = 6): Promise<RecommendationItem[]> {
    return this.getDestinationRecommendations(limit);
  },

  /**
   * Submit feedback (Helpful, Not Relevant, Dismiss) on a recommendation.
   */
  async submitFeedback(payload: SubmitFeedbackPayload): Promise<void> {
    try {
      await apiClient.post<ApiResponse<any>>('/v1/recommendations/feedback', payload);
    } catch (err) {
      console.error('Failed to submit recommendation feedback', err);
    }
  },

  /**
   * Get current user's inferred travel preference profile.
   */
  async getUserPreferences(): Promise<UserPreferenceProfile | null> {
    try {
      const response = await apiClient.get<ApiResponse<UserPreferenceProfile>>('/v1/recommendations/preferences');
      return response.data.data || null;
    } catch {
      return null;
    }
  },

  /**
   * Get recent user activity history.
   */
  async getUserHistory(limit: number = 20): Promise<any[]> {
    try {
      const response = await apiClient.get<ApiResponse<any[]>>('/v1/recommendations/history', {
        params: { limit },
      });
      return response.data.data || [];
    } catch {
      return [];
    }
  },

  /**
   * Track user activity for personalized recommendations.
   */
  async trackActivity(payload: TrackActivityPayload): Promise<void> {
    try {
      await apiClient.post<ApiResponse<void>>('/v1/recommendations/track', payload);
    } catch {
      // Background activity tracking failure shouldn't disrupt UI
    }
  },
};

File: backend/src/main/java/com/smarttravel/modules/recommendation/service/RecommendationService.java
package com.smarttravel.modules.recommendation.service;

import com.smarttravel.modules.recommendation.dto.RecommendationItem;
import com.smarttravel.modules.recommendation.dto.UserPreferenceProfileDto;
import com.smarttravel.modules.recommendation.model.RecommendationFeedback;
import com.smarttravel.modules.recommendation.model.RecommendationFeedbackType;
import com.smarttravel.modules.recommendation.model.UserActivity;
import com.smarttravel.modules.recommendation.model.UserActivityType;

import java.util.List;
import java.util.Map;

/**
 * Service for tracking user activity, recording feedback, and generating explainable personalized recommendations.
 */
public interface RecommendationService {

    /**
     * Record a user activity event.
     */
    void trackActivity(String userId, UserActivityType type, String targetId,
                       String targetType, Map<String, Object> metadata);

    /**
     * Get personalized recommendations for a user (mixed flights, hotels, and destinations).
     * Hybrid: content-based (28%) + activity-based (22%) + collaborative (20%) + preference (15%) + popularity (15%) + feedback tuning.
     */
    List<RecommendationItem> getRecommendations(String userId, int limit);

    /**
     * Get contextual recommendations with destination and context awareness.
     */
    List<RecommendationItem> getRecommendations(String userId, String context, String destination, int limit);

    /**
     * Get recommended flights for a user.
     */
    List<RecommendationItem> getFlightRecommendations(String userId, int limit);

    /**
     * Get recommended hotels for a user.
     */
    List<RecommendationItem> getHotelRecommendations(String userId, int limit);

    /**
     * Get popular & personalized destinations for a user.
     */
    List<RecommendationItem> getPopularDestinations(int limit);

    /**
     * Get destination recommendations tailored to user style & categories.
     */
    List<RecommendationItem> getDestinationRecommendations(String userId, int limit);

    /**
     * Record user feedback (Helpful, Not Relevant, Dismiss) on a recommendation.
     */
    RecommendationFeedback recordFeedback(String userId, String targetId, String targetType,
                                          RecommendationFeedbackType feedbackType, String reasonCode, String category);

    /**
     * Synthesize and retrieve the user's inferred travel preference profile.
     */
    UserPreferenceProfileDto getUserPreferenceProfile(String userId);

    /**
     * Get recent activity history for a user.
     */
    List<UserActivity> getUserActivityHistory(String userId, int limit);
}
